lib/Dialect/FIRRTL/FIRRTLTypes.cpp

```diff
@@ -1065,6 +1115,161 @@ std::pair<uint64_t, bool> FVectorType::rootChildFieldID(uint64_t fieldID,
                         fieldID >= childRoot && fieldID <= rangeEnd);
 }
 
+//===----------------------------------------------------------------------===//
+// Enum Type
+//===----------------------------------------------------------------------===//
+
+struct circt::firrtl::detail::EnumTypeStorage : mlir::TypeStorage {
+  using KeyTy = ArrayRef<EnumType::EnumElement>;
+
+  EnumTypeStorage(KeyTy elements) : elements(elements.begin(), elements.end()) {
+    RecursiveTypeProperties props{true, false, false};
+    uint64_t fieldID = 0;
+    fieldIDs.reserve(elements.size());
+    for (auto &element : elements) {
+      auto type = element.type;
+      auto eltInfo = type.getRecursiveTypeProperties();
+      props.isPassive &= eltInfo.isPassive;
+      props.containsAnalog |= eltInfo.containsAnalog;
+      props.hasUninferredWidth |= eltInfo.hasUninferredWidth;
+      fieldID += 1;
+      fieldIDs.push_back(fieldID);
+      // Increment the field ID for the next field by the number of subfields.
+      fieldID += type.getMaxFieldID();
+    }
+    maxFieldID = fieldID;
+    recProps = props;
+    assert(!props.containsAnalog && props.isPassive);
+  }
+
+  bool operator==(const KeyTy &key) const { return key == KeyTy(elements); }
+
+  static llvm::hash_code hashKey(const KeyTy &key) {
+    return llvm::hash_combine_range(key.begin(), key.end());
+  }
+
+  static EnumTypeStorage *construct(TypeStorageAllocator &allocator,
+                                    KeyTy key) {
+    return new (allocator.allocate<EnumTypeStorage>()) EnumTypeStorage(key);
+  }
+
+  SmallVector<EnumType::EnumElement, 4> elements;
+  SmallVector<uint64_t, 4> fieldIDs;
+  uint64_t maxFieldID;
+
+  RecursiveTypeProperties recProps;
+};
+
+auto EnumType::getElements() const -> ArrayRef<EnumElement> {
+  return getImpl()->elements;
+}
```

Comment:
I know this was c/p from BundleType, but we should probably use the standard function syntax.  Also interestingly, looks like the prototype for this is declared, but I guess you have to write the definition yourself since we aren't automatically generating the storage class.